iina/FFmpegController.m

```diff
@@ -0,0 +1,266 @@
+//
+//  FFmpegController.m
+//  iina
+//
+//  Created by lhc on 9/6/2017.
+//  Copyright © 2017 lhc. All rights reserved.
+//
+
+#import "FFmpegController.h"
+#import <Cocoa/Cocoa.h>
+
+#import <libavcodec/avcodec.h>
+#import <libavformat/avformat.h>
+#import <libswscale/swscale.h>
+#import <libavutil/imgutils.h>
+
+#define THUMB_COUNT_DEFAULT 100
+#define THUMB_WIDTH 240
+
+#define CHECK_NOTNULL(ptr,msg) if (ptr == NULL) {\
+NSLog(@"Error when getting thumbnails: %@", msg);\
+return -1;\
+}
+
+#define CHECK_SUCCESS(ret,msg) if (ret < 0) {\
+NSLog(@"Error when getting thumbnails: %@ (%d)", msg, ret);\
+return -1;\
+}
+
+@implementation FFThumbnail
+
+@end
+
+
+@interface FFmpegController () {
+  NSMutableArray<FFThumbnail *> *_thumbnails;
+  NSMutableSet *_addedTimestamps;
+  NSOperationQueue *_queue;
+}
+
+- (int)getPeeksForFile:(NSString *)file;
+- (void)sameThumbnail:(AVFrame *)pFrame width:(int)width height:(int)height index:(int)index realTime:(int)second;
+
+@end
+
+
+@implementation FFmpegController
+
+- (instancetype)init
+{
+  self = [super init];
+  if (self) {
+    self.thumbnailCount = THUMB_COUNT_DEFAULT;
+    _thumbnails = [[NSMutableArray alloc] init];
+    _addedTimestamps = [[NSMutableSet alloc] init];
+    _queue = [[NSOperationQueue alloc] init];
+  }
+  return self;
+}
+
+
+- (void)generateThumbnailForFile:(NSString *)file
+{
+  [_queue cancelAllOperations];
+  [_queue addOperationWithBlock: ^(){
+    int success = [self getPeeksForFile:file];
+    if (self.delegate) {
+      [self.delegate didGeneratedThumbnails:[NSArray arrayWithArray:_thumbnails] withSuccess:(success < 0 ? NO : YES)];
+    }
+  }];
+}
+
+
+- (int)getPeeksForFile:(NSString *)file
+{
+  int i, ret;
```

Comment:
Oh, I just don't want to write them together :stuck_out_tongue: (especially for function calls that spans multiple lines).